cirq/google/api/v2/program.proto

```diff
@@ -198,3 +216,81 @@ message ArgFunction {
   // The arguments to the function.
   repeated Arg args = 2;
 }
+
+// An operation that describes a circuit.
+message CircuitOperation {
+  // The index of the circuit in the top-level constant table.
+  int32 circuit_constant_index = 1;
+
+  // Specifier for repetitions of the circuit, including number of reps and
+  // repetition IDs.
+  RepetitionSpec repetition_spec = 2;
+
+  // Map of qubits in the circuit to the qubits used in this operation.
+  QubitMapping qubit_map = 3;
+
+  // Map of measurement keys in the circuit to the measurement keys used in
+  // this operation.
+  MeasurementKeyMapping measurement_key_map = 4;
+
+  // Map of args in the circuit to the values (or replacement args) used in
+  // this operation.
+  ArgMapping arg_map = 5;
+}
+
+// A description of the repetitions of a subcircuit.
+message RepetitionSpec {
+  // An ordered list of IDs for a sequence of repetitions.
+  message RepetitionIds {
+    repeated string ids = 1;
+  }
+
+  oneof rep_value {
+    // A list of unique IDs, one per repetition of the subcircuit.
+    RepetitionIds rep_ids = 1;
+
+    // An integer number of repetitions to perform.
+    int32 rep_count = 2;
+  }
+}
+
+// A mapping of qubits from one value to another.
+message QubitMapping {
+  // Indicates that qubit "first" should be replaced with "second".
```

Comment:
I wonder if first and second should be replaced with more descriptive terms.  Perhaps, source and target?  Or initial and mapped qubits?

Same goes for below.

Reply:
Done.

Reply:
I've been thinking of this as "inner" and "outer", where "inner" refers to the identifier inside the circuit operation, and "outer" refers to the identifier in the circuit in which that op is embedded.